### ISSUE TITLE
Sliding Window should be disabled if set to None or if >= seq length

### ISSUE THREAD

**Issue by DreamGenX** [axolotl-ai-cloud/axolotl#1213]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

Sliding window should be disabled when it's set to None, or when sliding window >= max seq length.

This is how it works in HuggingFace: https://github.com/huggingface/transformers/blob/8eb74c1c8961e3dc8549bb1a76463c7658a63d43/src/transformers/models/mistral/modeling_mistral.py#L380

This is for example used by the latest instruct model from Mistral AI: https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.2/blob/main/config.json#L19

### Current behaviour

Currently, sliding_window cannot be set to null / None, if so, this will fail: https://github.com/OpenAccess-AI-Collective/axolotl/blob/74c72ca5eb2e47345de6ed2422f48cf815d246ec/src/axolotl/monkeypatch/mistral_attn_hijack_flash.py#L102

Several users ran into this issue already when trying to fine tune instruct v0.2:

https://github.com/OpenAccess-AI-Collective/axolotl/issues/969
https://github.com/OpenAccess-AI-Collective/axolotl/issues/1047

### Steps to reproduce

Set

```
model_config:
  sliding_window: null
```

In model config while using a Mistral base model.

### Config yaml

```
model_config:
  sliding_window: null
```

### Possible solution

We could add `and sliding_window is not None` here: https://github.com/OpenAccess-AI-Collective/axolotl/blob/74c72ca5eb2e47345de6ed2422f48cf815d246ec/src/axolotl/monkeypatch/mistral_attn_hijack_flash.py#L102

But I wonder if this might have some other effects.

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

Using latest docker image

### axolotl branch-commit

Using latest docker image

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by DreamGenX** [comment]
It looks like axolotl already intends to disable sliding window here: https://github.com/OpenAccess-AI-Collective/axolotl/blob/74c72ca5eb2e47345de6ed2422f48cf815d246ec/src/axolotl/monkeypatch/mistral_attn_hijack_flash.py#L153

But even if use_sliding_windows is False, we end up calling `_prepare_decoder_attention_mask`.

There's also another bug, the code is:

```py
    use_sliding_windows = (
        hasattr(self.config, "sliding_window") is not None
        and kv_seq_len > self.config.sliding_window
    )
```

But I think the intention was to use `getattr` instead of `hasattr`. If you use `hasattr` the code will fail if sliding_window is set to None, as you can compare int and None (since hasattr will of course return true).

So my proposed fix is to change `hasattr` to `getattr` and to short-circuit `_prepare_decoder_attention_mask` is `sliding_window` is None.

**Comment by casper-hansen** [comment]
Yeah, it seems Mistral introduced `sliding_window=None` and caused some of axolotl's code to throw errors because of it. This should be a simple fix for the condition as you showed @DreamGenX.

**Comment by NanoCode012** [comment]
Thanks for this. I will link my original issue to yours.
